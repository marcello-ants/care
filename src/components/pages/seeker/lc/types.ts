import { GeneralProviderProfile } from '@/types/seeker';

export interface ProviderProfile extends GeneralProviderProfile {
  numberOfReviews: number;
  qualities?: ProviderQualities;
  services?: ProviderServices;
  firstName: string;
}

export interface ProviderQualities {
  certifiedNursingAssistant?: boolean;
  certifiedRegisterNurse?: boolean;
  certifiedTeacher?: boolean;
  childDevelopmentAssociate?: boolean;
  comfortableWithPets?: boolean;
  cprTrained?: boolean;
  crn?: boolean;
  doesNotSmoke?: boolean;
  doula?: boolean;
  earlyChildDevelopmentCoursework?: boolean;
  earlyChildhoodEducation?: boolean;
  firstAidTraining?: boolean;
  nafccCertified?: boolean;
  ownTransportation?: boolean;
  specialNeedsCare?: boolean;
  trustlineCertifiedCalifornia?: boolean;
}

export interface ProviderServices {
  carpooling?: boolean;
  craftAssistance?: boolean;
  errands?: boolean;
  groceryShopping?: boolean;
  laundryAssistance?: boolean;
  lightHousekeeping?: boolean;
  mealPreparation?: boolean;
  swimmingSupervision?: boolean;
  travel?: boolean;
}
