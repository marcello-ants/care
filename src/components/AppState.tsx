import { useRouter } from 'next/router';
import { createContext, useReducer, useContext, ReactNode, useEffect } from 'react';
import { debounce } from 'lodash-es';
import { configureScope, captureException, Severity } from '@sentry/nextjs';
import {
  initialAppState,
  rootReducer,
  isStateVersionCompatible,
  doesStateHaveDefaultKeys,
  validateState,
} from '@/state';
import patchAppState from '@/state/patch';
import { initialState as flowInitialState } from '@/state/flow';
import logger from '@/lib/clientLogger';
import { ClientLogger } from '@care/utils';
import getVertical from '@/utilities/verticalUtils';
import { SessionStorageHelper } from '@/utilities/SessionStorageHelper';
import { trackPageViewInSiftScience } from '@/utilities/siftScienceHelper';
import { AppAction, AppDispatch, AppState } from '@/types/app';
import { StateCheckingResponse } from '@/types/common';
import {
  LOCAL_STORAGE_STATE_KEY,
  CZEN_BASE_PATH,
  FLOWS,
  POST_A_JOB_ROUTES,
  CLIENT_FEATURE_FLAGS,
} from '@/constants';
import { useFeatureFlags } from '@/components/FeatureFlagsContext';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';

type AppStateProviderProps = {
  children: ReactNode;
  initialStateOverride?: AppState;
  czenJSessionId?: string | undefined;
  referrerCookie?: string | undefined | null;
  userHasAccount?: boolean;
  memberUuid?: string;
};

function getStateFromStorageOrDefault(defaultState: AppState) {
  let stateFromStorageJson;
  const sessionStorageAvailable = SessionStorageHelper.isAvailable();

  if (sessionStorageAvailable) {
    stateFromStorageJson = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
  } else {
    logger.warn({ event: 'sessionStorageUnavailable' });
  }

  if (stateFromStorageJson) {
    const stateFromStorage = JSON.parse(stateFromStorageJson) as AppState;
    try {
      const patchedState = patchAppState(stateFromStorage, defaultState);

      if (isStateVersionCompatible(patchedState) && doesStateHaveDefaultKeys(patchedState)) {
        return patchedState;
      }
      logger.warn({ event: 'invalidStateFromLocalStorage', state: stateFromStorageJson });
    } catch (e) {
      captureException(e, { level: Severity.Warning });
    }
  }
  return defaultState;
}

function persistState(state: AppState) {
  SessionStorageHelper.setItem(LOCAL_STORAGE_STATE_KEY, JSON.stringify(state));
}

const debouncePersistState = debounce(persistState, 500, {
  leading: true,
  trailing: true,
  maxWait: 1000,
});

function reducerWrapper(state: AppState, action: AppAction) {
  const nextState = rootReducer(state, action);
  debouncePersistState(nextState);
  return nextState;
}

export function redirectIfInvalidState(state: AppState, pathname: string): StateCheckingResponse {
  if (
    typeof Intl === 'undefined' ||
    typeof Intl.NumberFormat === 'undefined' ||
    typeof Intl.DateTimeFormat === 'undefined'
  ) {
    // TODO: Remove this with SC-214
    logger.error({ event: 'IntlUnavailable', source: pathname, invalidState: state });
    return {
      redirectCalled: true,
      readableMessage: 'Safari 9 and lower is not currently supported',
    };
  }

  const isStateValid = validateState(state, pathname);
  const redirect = !isStateValid;

  if (redirect) {
    SessionStorageHelper.removeItem(LOCAL_STORAGE_STATE_KEY);

    window.location.assign(CZEN_BASE_PATH);
  }

  return { redirectCalled: redirect };
}

export function getFlowName(pathname: string, oldFlowName?: string): string {
  let newFlowName = Object.values(FLOWS).find((flow) => pathname.includes(flow.path))?.name;

  if (!newFlowName) {
    // TODO: this shouldn't be needed after SC-123 is implemented
    const isPostAJob = Object.values(POST_A_JOB_ROUTES).some((route) => pathname.includes(route));
    if (isPostAJob) {
      newFlowName = 'SEEKER_POST_A_JOB';
    }
  }

  // Don't allow shifting from in-facility to seeker flow name
  if (newFlowName === FLOWS.SEEKER.name && oldFlowName === FLOWS.SEEKER_IN_FACILITY.name) {
    return FLOWS.SEEKER_IN_FACILITY.name;
  }

  return newFlowName ?? '';
}

function useFlowDetection(dispatch: AppDispatch, oldFlowName: string) {
  const { pathname } = useRouter();
  const newFlowName = getFlowName(pathname, oldFlowName);
  const vertical = getVertical();
  ClientLogger.setLoggerStatics({ flow: newFlowName, vertical });

  useEffect(() => {
    dispatch({ type: 'setFlowName', flowName: newFlowName });

    configureScope((scope) => {
      scope.setTags({
        flow: newFlowName,
        vertical,
      });
    });
  }, [newFlowName]);
}

const AppStateContext = createContext<AppState>(initialAppState);
const AppDispatchContext = createContext<AppDispatch>(() =>
  // eslint-disable-next-line no-console
  console.warn(
    'Calling default dispatch is a no-op! You are probably consuming app dispatch in a component above AppStateProvider'
  )
);

export const AppStateProvider = ({
  children,
  initialStateOverride,
  czenJSessionId,
  referrerCookie,
  userHasAccount,
  memberUuid,
}: AppStateProviderProps) => {
  const hydratedInitialAppState = {
    ...initialAppState,
    flow: {
      ...flowInitialState,
      czenJSessionId,
      referrerCookie,
      userHasAccount,
      memberUuid,
    },
  };
  const featureFlags = useFeatureFlags();
  const sessionStorageRetrievalSuccessEvaluation =
    featureFlags?.flags[CLIENT_FEATURE_FLAGS.SESSION_STORAGE_RETRIEVAL_SUCCESS];
  const shouldMeasureSessionStorageRetrievalSuccess =
    sessionStorageRetrievalSuccessEvaluation?.variationIndex === 1;

  const router = useRouter();

  const logSessionStorageError = () => {
    logger.info({ event: 'sessionStorageStateUnavailable', path: router.pathname });
    AnalyticsHelper.logEvent({
      name: 'sessionStorageStateUnavailable',
    });
  };
  const launchDarklyWrappedGetStateFromStorageOrDefault = (defaultState: AppState) => {
    if (shouldMeasureSessionStorageRetrievalSuccess) {
      try {
        const sessionStorageState = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
        if (sessionStorageState && sessionStorageState.length > 0) {
          const stateFromStorage = JSON.parse(sessionStorageState);
          if (stateFromStorage) {
            logger.info({ event: 'sessionStorageStateAvailable', path: router.pathname });
          } else {
            logSessionStorageError();
          }
        } else {
          logSessionStorageError();
        }
      } catch (_e) {
        logSessionStorageError();
      }
    }
    return getStateFromStorageOrDefault(defaultState);
  };

  const [state, dispatch] = useReducer(
    reducerWrapper,
    initialStateOverride ?? hydratedInitialAppState,
    launchDarklyWrappedGetStateFromStorageOrDefault
  );
  useFlowDetection(dispatch, state.flow.flowName);
  const routePath = router.asPath?.split('?')[0];
  const stateCheckResult = redirectIfInvalidState(state, routePath);
  if (stateCheckResult.redirectCalled) {
    if (stateCheckResult.readableMessage) {
      return (
        <>
          <p>Redirecting to Member Home Page...</p>
          <p>{stateCheckResult.readableMessage}</p>
        </>
      );
    }
    return <></>;
  }

  useEffect(() => {
    trackPageViewInSiftScience(state);
  }, [routePath]);

  return (
    <AppDispatchContext.Provider value={dispatch}>
      <AppStateContext.Provider value={state}>{children}</AppStateContext.Provider>
    </AppDispatchContext.Provider>
  );
};

AppStateProvider.defaultProps = {
  initialStateOverride: undefined,
  czenJSessionId: undefined,
  referrerCookie: undefined,
  userHasAccount: false,
  memberUuid: undefined,
};

export const useAppState = () => useContext(AppStateContext);
export const useAppDispatch = () => useContext(AppDispatchContext);

export const useSeekerState = () => useContext(AppStateContext).seeker;
export const useProviderState = () => useContext(AppStateContext).provider;
export const useProviderCCState = () => useContext(AppStateContext).providerCC;
export const useFlowState = () => useContext(AppStateContext).flow;
export const useSeekerCCState = () => useContext(AppStateContext).seekerCC;
export const useSeekerHKState = () => useContext(AppStateContext).seekerHK;
export const useSeekerPCState = () => useContext(AppStateContext).seekerPC;
export const useSeekerTUState = () => useContext(AppStateContext).seekerTU;
export const useLtcgState = () => useContext(AppStateContext).ltcg;
