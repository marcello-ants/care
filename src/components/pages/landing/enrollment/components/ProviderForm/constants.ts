export const PROVIDER_SERVICES_DESKTOP = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const PROVIDER_SERVICES_MOBILE = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const DROPOFF_URLS = {
  CHILD_CARE: '/swf/ccOptimizedFlow',
  SENIOR_CARE: '/app/enrollment/provider/sc ',
  PET_CARE: '/swf/pcOptimizedFlow',
  HOUSEKEEPING: '/swf/hkOptimizedFlow',
  TUTORING: '/swf/tuOptimizedFlow',
  SPECIAL_NEEDS: '/swf/snOptimizedFlow',
  GENERIC: '/swf/cgOptimizedFlow',
  AU_PAIR: '',
  CARE_GIGS: '',
};
