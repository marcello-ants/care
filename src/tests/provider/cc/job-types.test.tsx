import React from 'react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { render, act, RenderResult, waitFor, screen, fireEvent } from '@testing-library/react';
import { useRouter } from 'next/router';
import { AppStateProvider } from '@/components/AppState';
import { initialAppState } from '@/state';
import { initialRecurringState } from '@/types/common';
import JobTypesPage from '@/pages/provider/cc/job-types';
import {
  ServiceType,
  JobInterestUpdateSource,
  JobHoursUnit,
  Language,
  ChildCareAgeGroups,
} from '@/__generated__/globalTypes';
import { PROVIDER_JOB_INTEREST_UPDATE, GET_DEFAULT_JOB_WAGES } from '@/components/request/GQL';
import { ThemeProvider } from '@material-ui/core';
import { theme } from '@care/material-ui-theme';
import { GraphQLError } from 'graphql';
import { PROVIDER_CHILD_CARE_ROUTES } from '@/constants';

jest.mock('next/router', () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

const zipcode = '01234';

const getDefaultJobWagesMock = {
  request: {
    query: GET_DEFAULT_JOB_WAGES,
    variables: {
      serviceType: ServiceType.CHILD_CARE,
      zipcode,
    },
  },
  result: {
    data: {
      getJobWages: {
        legalMinimum: { amount: '14', __typename: 'Money' },
        defaultMaxWage: { amount: '19', __typename: 'Money' },
        defaultMinWage: { amount: '14', __typename: 'Money' },
        maxAllowed: { amount: '50', __typename: 'Money' },
        __typename: 'WagesPayload',
      },
    },
  },
};

const jobTypesOneTimeUpdateSuccessMock = {
  request: {
    query: PROVIDER_JOB_INTEREST_UPDATE,
    variables: {
      input: {
        source: JobInterestUpdateSource.ENROLLMENT,
        oneTimeJobInterest: {
          jobHours: {
            minimum: 7,
            unit: JobHoursUnit.PER_JOB,
          },
          childCareJobRates: [
            {
              numberOfChildren: 1,
              hourlyRate: {
                currencyCode: 'USD',
                amount: '14',
              },
            },
          ],
        },
        recurringJobInterest: null,
        serviceType: ServiceType.CHILD_CARE,
      },
    },
  },
  result: {
    data: {
      providerJobInterestUpdate: { __typename: 'ProviderJobInterestUpdateSuccess', dummy: null },
    },
  },
};

const jobTypesOneTimeUpdateFailedMock = {
  request: {
    query: PROVIDER_JOB_INTEREST_UPDATE,
    variables: {
      input: {
        source: JobInterestUpdateSource.ENROLLMENT,
        oneTimeJobInterest: {
          jobHours: {
            minimum: 7,
            unit: JobHoursUnit.PER_JOB,
          },
          childCareJobRates: [
            {
              numberOfChildren: 1,
              hourlyRate: {
                currencyCode: 'USD',
                amount: '14',
              },
            },
          ],
        },
        recurringJobInterest: null,
        serviceType: ServiceType.CHILD_CARE,
      },
    },
  },
  result: { errors: [new GraphQLError('error')] },
};

const jobTypesRecurringUpdateSuccessMock = {
  request: {
    query: PROVIDER_JOB_INTEREST_UPDATE,
    variables: {
      input: {
        source: JobInterestUpdateSource.ENROLLMENT,
        oneTimeJobInterest: null,
        recurringJobInterest: {
          jobHours: {
            minimum: 7,
            maximum: 9,
            unit: JobHoursUnit.PER_WEEK,
          },
          jobRate: {
            maximum: {
              amount: '19',
              currencyCode: 'USD',
            },
            minimum: {
              amount: '14',
              currencyCode: 'USD',
            },
          },
        },
        serviceType: ServiceType.CHILD_CARE,
      },
    },
  },
  result: {
    data: {
      providerJobInterestUpdate: { __typename: 'ProviderJobInterestUpdateSuccess', dummy: null },
    },
  },
};

const jobTypesRecurringUpdateWithoutMaximumSuccessMock = {
  request: {
    query: PROVIDER_JOB_INTEREST_UPDATE,
    variables: {
      input: {
        source: JobInterestUpdateSource.ENROLLMENT,
        oneTimeJobInterest: null,
        recurringJobInterest: {
          jobHours: {
            minimum: 7,
            maximum: undefined,
            unit: JobHoursUnit.PER_WEEK,
          },
          jobRate: {
            maximum: {
              amount: '19',
              currencyCode: 'USD',
            },
            minimum: {
              amount: '14',
              currencyCode: 'USD',
            },
          },
        },
        serviceType: ServiceType.CHILD_CARE,
      },
    },
  },
  result: {
    data: {
      providerJobInterestUpdate: { __typename: 'ProviderJobInterestUpdateSuccess', dummy: null },
    },
  },
};

describe('Job Types', () => {
  let renderResult: RenderResult;
  let mockRouter: any | null = null;

  async function renderComponent(mocks: MockedResponse<Record<string, any>>[]) {
    const state = {
      ...initialAppState,
      providerCC: {
        city: '',
        state: '',
        zipcode,
        distance: 15,
        lat: null,
        lng: null,
        recurringJob: false,
        oneTime: false,
        defaultRatePerChild: [0],
        defaultRecurringHourlyRate: [0, 0],
        min: '',
        max: '',
        hours: '',
        initialAccountCreationAttempted: false,
        caregiverPreferencesPersisted: false,
        firstName: '',
        lastName: '',
        headline: '',
        bio: '',
        hourlyRate: 14,
        numberOfJobsNear: 0,
        jobTypes: [],
        jobTypesSchedules: {},
        recurring: { ...initialRecurringState.recurring },
        languages: [Language.ENGLISH],
        education: 'SOME_HIGH_SCHOOL',
        selectedAboutMe: ['smokes'],
        selectedHelpWith: [],
        selectedSkills: [],
        experienceYears: 1,
        numberOfChildren: 1,
        ageGroups: [ChildCareAgeGroups.EARLY_SCHOOL],
        careForSickChild: false,
        covidVaccinated: '',
        freeGated: false,
        callFreeGatedMutation: true,
        callProviderNameUpdateMutation: true,
        textAnalysisValidationId: '',
      },
    };

    renderResult = render(
      <ThemeProvider theme={theme}>
        <MockedProvider mocks={mocks} addTypename>
          <AppStateProvider initialStateOverride={state}>
            <JobTypesPage />
          </AppStateProvider>
        </MockedProvider>
      </ThemeProvider>
    );

    await act(() => new Promise<undefined>((resolve) => setTimeout(resolve, 0)));
  }

  beforeEach(() => {
    mockRouter = {
      push: jest.fn(), // the component uses `router.push` only
      query: { param: [] },
      pathname: '/provider/cc/job-types',
    };
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
  });

  afterEach(() => {
    // cleanup on exiting
    mockRouter = null;
  });

  it('should match snapshot', async () => {
    await renderComponent([getDefaultJobWagesMock]);
    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should disable submit button by default', async () => {
    await renderComponent([getDefaultJobWagesMock]);

    const btn = screen.getByRole('button', { name: 'Next' });
    await waitFor(() => expect(btn).toBeDisabled());
  });

  it('should enable submit button when user selects one time job and enters hours', async () => {
    await renderComponent([getDefaultJobWagesMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'One-time jobs Date nights, backup sitter, and special occasions. Pick up occasional jobs based on your preferred schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const hoursInput = screen.getByPlaceholderText('Hours');
    fireEvent.change(hoursInput, { target: { value: '7' } });

    const btn = screen.getByRole('button', { name: 'Next' });
    await waitFor(() => expect(btn).toBeEnabled());
  });

  it('should be able to add several rates for children for one time job', async () => {
    await renderComponent([getDefaultJobWagesMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'One-time jobs Date nights, backup sitter, and special occasions. Pick up occasional jobs based on your preferred schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const addChild = await screen.findByText('Add another child');
    fireEvent.click(addChild);

    const first = await screen.findByText('1 child');
    const second = await screen.findByText('2 children');
    expect(first).toBeInTheDocument();
    expect(second).toBeInTheDocument();
  });

  it('should be able to remove rate for one time job', async () => {
    await renderComponent([getDefaultJobWagesMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'One-time jobs Date nights, backup sitter, and special occasions. Pick up occasional jobs based on your preferred schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const addChild = await screen.findByText('Add another child');
    fireEvent.click(addChild);

    const removeChildRate = await screen.findByTestId('remove-rate');
    fireEvent.click(removeChildRate);

    const first = screen.getByText('1 child');
    const second = screen.queryByText('2 children');
    expect(first).toBeInTheDocument();
    expect(second).not.toBeInTheDocument();
  });

  it('should be able to add up to 4 rates for children for one time job', async () => {
    await renderComponent([getDefaultJobWagesMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'One-time jobs Date nights, backup sitter, and special occasions. Pick up occasional jobs based on your preferred schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const addChild = screen.getByText('Add another child');
    fireEvent.click(addChild);
    fireEvent.click(addChild);
    fireEvent.click(addChild);

    expect(screen.queryByText('Add another child')).not.toBeInTheDocument();
  });

  it('should be able to submit changes for one time job', async () => {
    await renderComponent([getDefaultJobWagesMock, jobTypesOneTimeUpdateSuccessMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'One-time jobs Date nights, backup sitter, and special occasions. Pick up occasional jobs based on your preferred schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const hoursInput = screen.getByPlaceholderText('Hours');
    fireEvent.change(hoursInput, { target: { value: '7' } });

    const btn = screen.getByRole('button', { name: 'Next' });
    fireEvent.click(btn);

    await waitFor(() =>
      expect(mockRouter.push).toHaveBeenCalledWith(PROVIDER_CHILD_CARE_ROUTES.PROFILE)
    );
  });

  it('should be able to redirect after failing mutation', async () => {
    await renderComponent([getDefaultJobWagesMock, jobTypesOneTimeUpdateFailedMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'One-time jobs Date nights, backup sitter, and special occasions. Pick up occasional jobs based on your preferred schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const hoursInput = screen.getByPlaceholderText('Hours');
    fireEvent.change(hoursInput, { target: { value: '7' } });

    const btn = screen.getByRole('button', { name: 'Next' });
    fireEvent.click(btn);
    await act(() => new Promise<undefined>((resolve) => setTimeout(resolve, 0)));

    const gotItBtn = await screen.findByRole('button', { name: 'Got it' });
    await waitFor(() => expect(gotItBtn).toBeInTheDocument());
    fireEvent.click(gotItBtn);

    await waitFor(() => expect(mockRouter.push).toHaveBeenCalledTimes(1));
  });

  it('should be able to redirect after failing mutation to Profile page', async () => {
    await renderComponent([getDefaultJobWagesMock, jobTypesOneTimeUpdateFailedMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'One-time jobs Date nights, backup sitter, and special occasions. Pick up occasional jobs based on your preferred schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const hoursInput = screen.getByPlaceholderText('Hours');
    fireEvent.change(hoursInput, { target: { value: '7' } });

    const btn = screen.getByRole('button', { name: 'Next' });
    fireEvent.click(btn);
    await act(() => new Promise<undefined>((resolve) => setTimeout(resolve, 0)));

    const gotItBtn = await screen.findByRole('button', { name: 'Got it' });
    await waitFor(() => expect(gotItBtn).toBeInTheDocument());
    fireEvent.click(gotItBtn);

    await waitFor(() =>
      expect(mockRouter.push).toHaveBeenCalledWith(PROVIDER_CHILD_CARE_ROUTES.PROFILE)
    );
  });

  it('should enable submit button when user selects recurring job and enters min/max hours', async () => {
    await renderComponent([getDefaultJobWagesMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'Recurring jobs Nanny, after-school and summer jobs. Work a regular, full, or part-time schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);
    await act(() => new Promise<undefined>((resolve) => setTimeout(resolve, 0)));

    const minInput = screen.getByPlaceholderText('Min');
    fireEvent.change(minInput, { target: { value: '7' } });

    const maxInput = screen.getByPlaceholderText('Max');
    fireEvent.change(maxInput, { target: { value: '9' } });

    const btn = screen.getByRole('button', { name: 'Next' });
    await waitFor(() => expect(btn).toBeEnabled());
  });

  it('should be able to submit recurring job', async () => {
    await renderComponent([getDefaultJobWagesMock, jobTypesRecurringUpdateSuccessMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'Recurring jobs Nanny, after-school and summer jobs. Work a regular, full, or part-time schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const minInput = screen.getByPlaceholderText('Min');
    fireEvent.change(minInput, { target: { value: '7' } });

    const maxInput = screen.getByPlaceholderText('Max');
    fireEvent.change(maxInput, { target: { value: '9' } });

    const btn = screen.getByRole('button', { name: 'Next' });
    fireEvent.click(btn);

    await waitFor(() =>
      expect(mockRouter.push).toHaveBeenCalledWith(PROVIDER_CHILD_CARE_ROUTES.AVAILABILITY)
    );
  });

  it('should be able to submit recurring job without maximum', async () => {
    await renderComponent([
      getDefaultJobWagesMock,
      jobTypesRecurringUpdateWithoutMaximumSuccessMock,
    ]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'Recurring jobs Nanny, after-school and summer jobs. Work a regular, full, or part-time schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const minInput = screen.getByPlaceholderText('Min');
    fireEvent.change(minInput, { target: { value: '7' } });

    const btn = screen.getByRole('button', { name: 'Next' });
    fireEvent.click(btn);

    await waitFor(() => expect(mockRouter.push).toHaveBeenCalledTimes(1));
  });

  it('should disable submit button when user writes a wrong value for hours', async () => {
    await renderComponent([getDefaultJobWagesMock]);

    const oneTimeJobCheckbox = screen.getByRole('checkbox', {
      name: 'Recurring jobs Nanny, after-school and summer jobs. Work a regular, full, or part-time schedule.',
    });
    fireEvent.click(oneTimeJobCheckbox);

    const minInput = screen.getByPlaceholderText('Min');
    fireEvent.change(minInput, { target: { value: '7' } });

    const maxInput = screen.getByPlaceholderText('Max');
    fireEvent.change(maxInput, { target: { value: '5' } });

    const btn = screen.getByRole('button', { name: 'Next' });
    await waitFor(() => expect(btn).toBeDisabled());
  });
});
