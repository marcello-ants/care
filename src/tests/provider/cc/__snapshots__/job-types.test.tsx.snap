// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job Types should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-gridContainer-1 MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h2
        class="MuiTypography-root makeStyles-header-19 MuiTypography-h2"
      >
        My schedule works best with:
      </h2>
    </div>
    <div
      class="MuiGrid-root makeStyles-header-2 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <p
        class="MuiTypography-root makeStyles-disclaimer1-26 MuiTypography-body1 MuiTypography-colorTextSecondary"
      >
        Select all that apply
      </p>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <label
        class="MuiFormControlLabel-root makeStyles-root-34 MuiFormControlLabel-labelPlacementStart"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-35 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-38"
              data-indeterminate="false"
              name="recurring"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-root-65"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5.6.75h12.8a4.8 4.8 0 014.85 4.85v12.8a4.8 4.8 0 01-4.85 4.85H5.6A4.8 4.8 0 01.75 18.4V5.6A4.93 4.93 0 015.6.75z"
                fill="#fff"
              />
              <path
                d="M18.4 23.27H5.6A4.88 4.88 0 01.73 18.4V5.6A4.93 4.93 0 015.6.73h12.8a4.88 4.88 0 014.87 4.87v12.8a4.88 4.88 0 01-4.87 4.87zM5.6 1.73A3.92 3.92 0 001.73 5.6v12.8a3.88 3.88 0 003.87 3.87h12.8a3.88 3.88 0 003.87-3.87V5.6a3.88 3.88 0 00-3.87-3.87z"
                fill="#23405f"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 MuiTypography-colorPrimary"
          >
            Recurring jobs
          </h4>
          <p
            class="MuiTypography-root makeStyles-body3-23 MuiTypography-body1 MuiTypography-colorSecondary"
          >
            Nanny, after-school and summer jobs. Work a regular, full, or part-time schedule.
          </p>
        </span>
      </label>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="makeStyles-divider-40 makeStyles-is-hairline-41"
        data-testid="divider"
      />
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <label
        class="MuiFormControlLabel-root makeStyles-root-34 MuiFormControlLabel-labelPlacementStart"
      >
        <span
          aria-disabled="false"
          class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-35 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
        >
          <span
            class="MuiIconButton-label"
          >
            <input
              class="PrivateSwitchBase-input-38"
              data-indeterminate="false"
              name="oneTime"
              type="checkbox"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-root-66"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5.6.75h12.8a4.8 4.8 0 014.85 4.85v12.8a4.8 4.8 0 01-4.85 4.85H5.6A4.8 4.8 0 01.75 18.4V5.6A4.93 4.93 0 015.6.75z"
                fill="#fff"
              />
              <path
                d="M18.4 23.27H5.6A4.88 4.88 0 01.73 18.4V5.6A4.93 4.93 0 015.6.73h12.8a4.88 4.88 0 014.87 4.87v12.8a4.88 4.88 0 01-4.87 4.87zM5.6 1.73A3.92 3.92 0 001.73 5.6v12.8a3.88 3.88 0 003.87 3.87h12.8a3.88 3.88 0 003.87-3.87V5.6a3.88 3.88 0 00-3.87-3.87z"
                fill="#23405f"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 MuiTypography-colorPrimary"
          >
            One-time jobs
          </h4>
          <p
            class="MuiTypography-root makeStyles-body3-23 MuiTypography-body1 MuiTypography-colorSecondary"
          >
            Date nights, backup sitter, and special occasions. Pick up occasional jobs based on your preferred schedule.
          </p>
        </span>
      </label>
    </div>
    <div
      class="MuiGrid-root makeStyles-nextButtonContainer-6 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge Mui-disabled MuiButton-fullWidth Mui-disabled"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Next
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
